rs/bitcoin/kyt/src/fetch.rs

```diff
@@ -0,0 +1,266 @@
+use crate::blocklist_contains;
+use crate::state::{FetchGuardError, FetchTxStatus, FetchedTx, HttpGetTxError};
+use crate::types::{CheckTransactionError, CheckTransactionResponse};
+use bitcoin::{address::FromScriptError, Address, Network, Transaction};
+use futures::future::try_join_all;
+use ic_btc_interface::Txid;
+use std::convert::Infallible;
+
+#[cfg(test)]
+mod tests;
+
+pub fn get_tx_cycle_cost(max_response_bytes: u32) -> u128 {
+    // 1 KiB for request, max_response_bytes for response
+    49_140_000 + 1024 * 5_200 + 10_400 * (max_response_bytes as u128)
+}
+
+/// Caller of check_transaction must attach this amount of cycles with the call.
+pub const CHECK_TRANSACTION_CYCLES_REQUIRED: u128 = 40_000_000_000;
+
+/// One-time charge for every check_transaction call.
+pub const CHECK_TRANSACTION_CYCLES_SERVICE_FEE: u128 = 100_000_000;
+
+// The max_response_bytes is initially set to 4kB, and then
+// increased to 400kB if the initial size isn't enough.
+// - The maximum size of a standard non-taproot transaction is 400k vBytes.
+// - Taproot transactions could be as big as full block size (4MiB).
+// - Currently a subnet's maximum response size is only 2MiB.
+// - Transaction size between 400kB and 2MiB are also uncommon, we could
+//   handle them in the future if required.
+// - Transactions bigger than 2MiB are very rare, and we can't handle them.
+
+/// Initial max response bytes is 4kB
+pub const INITIAL_MAX_RESPONSE_BYTES: u32 = 4 * 1024;
+
+/// Retry max response bytes is 400kB
+pub const RETRY_MAX_RESPONSE_BYTES: u32 = 400 * 1024;
+
+pub enum FetchResult {
+    RetryWithBiggerBuffer,
+    Error(HttpGetTxError),
+    Fetched(FetchedTx),
+}
+
+pub enum TryFetchResult<F> {
+    Pending,
+    HighLoad,
+    Error(HttpGetTxError),
+    NotEnoughCycles,
+    Fetched(FetchedTx),
+    ToFetch(F),
+}
+
+/// Trait that abstracts over state operations.
+pub trait FetchState {
+    type FetchGuard;
+    fn new_fetch_guard(&self, txid: Txid) -> Result<Self::FetchGuard, FetchGuardError>;
+    fn get_fetch_status(&self, txid: Txid) -> Option<FetchTxStatus>;
+    fn set_fetch_status(&self, txid: Txid, status: FetchTxStatus);
+    fn set_fetched_address(&self, txid: Txid, index: usize, address: Address);
+}
```

Comment:
It's not yet clear to me why this is needed/beneficial. The unit tests IMO should use the productive logic regarding manipulating the canister state (read/write) to ensure that this logic is also tested via unit tests, but now they use a test implementation. Would it be possible to only keep the productive implementation (`KytCanisterState`)

Reply:
I think you are right. I only needed to mock the behavior of `new_fetch_guard` to make testing easier.

Just pushed a commit that removes `FetchState` and directly uses global state functions. Please have a look, thanks!

Reply:
Nice, thanks for the change!